src/appleseed/foundation/utility/unzipper.cpp

```diff
@@ -0,0 +1,217 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+
+// Interface header.
+#include "unzipper.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exception.h"
+#include "foundation/utility/minizip/unzip.h"
+
+// Boost headers.
+#include "boost/filesystem.hpp"
+
+// Standard headers.
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <sstream>
+
+using namespace std;
+namespace bf = boost::filesystem;
+
+namespace foundation
+{
+
+const int BUFFER_SIZE = 4096;
+
+const char* UnzipException::what() const throw() 
+{
+    stringstream exc_message;
+    exc_message << message;
+
+    if (err != 0)
+        exc_message << err;
+
+    return exc_message.str().c_str();
+}
+
+bool exists(const string& dirname) {
+    bf::path dir(dirname);
+    return bf::exists(dir);
+}
+
+bool make_dir(const string& dirname) 
+{
+    bf::path dir(dirname);
+    return bf::create_directories(dir);
+}
+
+int remove_dir(const string& dirname)
+{
+    bf::path dir(dirname);
+    return bf::remove_all(dir);
+}
+
+bool is_directory(const string& dirname) 
+{
+    // used own implementation of is_directory instead of boost implementation
+    // because this directory is not in filesystem, but in zipfile
+    return dirname[dirname.size() - 1] == '/';
+}
+
+void open_current_file(unzFile& zip_file) 
+{
+    const int err = unzOpenCurrentFile(zip_file);
+    if (err != UNZ_OK) 
+        throw UnzipException("Can't open file inside zip: ", err);
+}
+
+int read_chunk(unzFile& zip_file, char* buffer, const int chunk_size) {
+    const int err = unzReadCurrentFile(zip_file, buffer, chunk_size);
+
+    if (err == UNZ_ERRNO)
+        throw UnzipException("IO error while reading from zip");
+    if (err < 0)
+        throw UnzipException("zLib error with decompessing file: ", err);
+
+    return err;
+}
+
+void close_current_file(unzFile& zip_file) {
+    const int err = unzCloseCurrentFile(zip_file);
+
+    if (err == UNZ_CRCERROR)
+        throw UnzipException("CRC32 is not good");
+}
+
+string read_filename(unzFile& zip_file) 
+{
+    char filename_inzip[256];
+    unzGetCurrentFileInfo(zip_file, NULL, filename_inzip, sizeof(filename_inzip), NULL, 0, NULL, 0);
+
+    return filename_inzip;
+}
+
+string get_filepath(unzFile& zip_file, const string& unzipped_dir) {
+    string filename = read_filename(zip_file);
+    return (bf::path(unzipped_dir) / bf::path(filename)).string();
+}
+
+void extract_current_file(unzFile& zip_file, const string& unzipped_dir) 
+{
+    const string filepath = get_filepath(zip_file, unzipped_dir);
+
+    if (is_directory(filepath)) 
+    {
+        make_dir(filepath);
+        return;
+    } else 
+        open_current_file(zip_file);
+
+    fstream out(filepath.c_str(), ios_base::out | ios_base::binary);
+
+    char* buffer = new char[BUFFER_SIZE];
```

Comment:
This code will leak in case of exceptions (there might not be exceptions, but it's a good practice to make code exception-safe).

If `BUFFER_SIZE` is small enough, allocate `buffer` on the stack:

    char buffer[BUFFER_SIZE];

If it's too large to be allocated on the stack, use an `std::vector`.